packages/agent/src/agent/http/index.ts

```diff
@@ -407,63 +417,167 @@ export class HttpAgent implements Agent {
+          body,
+        }),
       );
-    }
 
-    const canister = typeof canisterId === 'string' ? Principal.fromText(canisterId) : canisterId;
-    const sender = id?.getPrincipal() || Principal.anonymous();
+      const queryResponse: QueryResponse = cbor.decode(await response.arrayBuffer());
 
-    const request: QueryRequest = {
-      request_type: ReadRequestType.Query,
-      canister_id: canister,
-      method_name: fields.methodName,
-      arg: fields.arg,
-      sender,
-      ingress_expiry: new Expiry(DEFAULT_INGRESS_EXPIRY_DELTA_IN_MSECS),
+      return {
+        ...queryResponse,
+        httpDetails: {
+          ok: response.ok,
+          status: response.status,
+          statusText: response.statusText,
+          headers: httpHeadersTransform(response.headers),
+        },
+        requestId,
+      };
     };
+    const queryPromise = new Promise<ApiQueryResponse>((resolve, reject) => {
+      makeQuery()
+        .then(response => {
+          resolve(response);
+        })
+        .catch(error => {
+          reject(error);
+        });
+    });
 
-    // TODO: remove this any. This can be a Signed or UnSigned request.
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    let transformedRequest: any = await this._transform({
-      request: {
-        method: 'POST',
-        headers: {
-          'Content-Type': 'application/cbor',
-          ...(this._credentials ? { Authorization: 'Basic ' + btoa(this._credentials) } : {}),
-        },
-      },
-      endpoint: Endpoint.Query,
-      body: request,
+    const subnetStatusPromise = new Promise<SubnetStatus | void>((resolve, reject) => {
+      if (!this.#verifyQuerySignatures) {
+        resolve();
+      }
+      const subnetStatus = this.#subnetKeys.get(canisterId.toString());
+      if (subnetStatus) {
+        resolve(subnetStatus);
+      } else {
+        this.fetchSubnetKeys(canisterId)
+          .then(response => {
+            resolve(response);
+          })
+          .catch(error => {
+            reject(error);
+          });
+      }
     });
+    const [query, subnetStatus] = await Promise.all([queryPromise, subnetStatusPromise]);
+    // Skip verification if the user has disabled it
+    if (!this.#verifyQuerySignatures) {
+      return query;
+    }
+    return this.#verifyQueryResponse(query, subnetStatus);
+  }
 
-    // Apply transform for identity.
-    transformedRequest = await id?.transformRequest(transformedRequest);
+  /**
+   * See https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-query for details on validation
+   * @param queryResponse - The response from the query
+   * @param subnetStatus - The subnet status, including all node keys
+   * @returns ApiQueryResponse
+   */
+  #verifyQueryResponse = (
+    queryResponse: ApiQueryResponse,
+    subnetStatus: SubnetStatus | void,
+  ): ApiQueryResponse => {
+    if (!subnetStatus || this.#verifyQuerySignatures === false) {
```

Comment:
It seems the only usage of `#verifyQueryResponse` is line 514. When would be case that `subnetStatus` is void?

Reply:
I'd suggest a re-read. VerifyQueryResponse is a new configuration option that defaults to true. SubnetStatus can be void if the developer deliberately opts out of verification, similar to what agent-rs is doing

Reply:
It seems `subnetStatusPromise` is always executed thus setting `verifyQuerySignatures` as false only skips verification but it still fetches node keys. The agent-rs code skips both.

Reply:
This is the same comment as given above.

IIUC, fetchSubnetKeys on line 518 is only executed if the check against #verifyQuerySignatures on line 511 fails.

That seems to ensure node keys are only fetched if #verifyQuerySignature is true.

Reply:
Also on the subnetStatus being void, I agree that #verifyQueryResponse shouldn't be called with void the calling code seems to assure that. It's still good to check that it is actually the case inside #verifyQueryResponse.

fetchSubnetKeys seems to ensure that subnetStatus is not void.

However, if for some reason #verifyQueryResponse was called with this.#verifyQuerySignatures === true but subnetStatus void, it would not do any verification and just return the queryResponse.

Shouldn't verification fail in such a situation?

Reply:
Yes. In fact, fixing this check fixed a number of false positives in unit tests that ought to have verification disabled